shelley/chain-and-ledger/executable-spec/test/Generator/Delegation.hs

```diff
@@ -171,3 +180,30 @@ genDCertRegPool :: KeyPairs -> VrfKeyPairs -> Gen (DCert, KeyPair)
 genDCertRegPool skeys vrfKeys = do
   (pps, poolKey) <- genStakePool skeys vrfKeys
   pure (RegPool pps, poolKey)
+
+-- | Generate a RetirePool along with the keypair which registered it.
+genRetirePool
+  :: KeyPairs
+  -> PParams
+  -> PState
+  -> Slot
+  -> Gen (Maybe (DCert, KeyPair))
+genRetirePool availableKeys pp pState slot =
+  if (null availableKeys || null poolHashKeys)
+     then pure Nothing
+     else (\keyHash epoch ->
+              Just (RetirePool keyHash epoch, findKeyPair keyHash))
+            <$> Gen.element poolHashKeys
+            <*> (Epoch <$> Gen.integral (Range.constant epochLow epochHigh))
+ where
+  stakePools = pState ^. (stPools . to (\(StakePools x) -> x))
+  poolHashKeys = Set.toList (Map.keysSet stakePools)
+  findKeyPair keyHash =
+    case find (\x -> hashKey (vKey (snd x)) == keyHash) availableKeys of
+      Nothing ->
+        error "genRetirePool: impossible: keyHash doesn't match availableKeys"
+      Just ks -> snd ks
+  Epoch cepoch = epochFromSlot slot
+  Epoch maxEpoch = pp ^. eMax
+  epochLow = cepoch + 1
+  epochHigh = cepoch + maxEpoch - 1
```

Comment:
I have modified this module in addition to `Generator.Delegation.QuickCheck`, for consistency. It seems inevitable that they will soon get out of sync, though, but hopefully it's not too long before the transition over to QC happens and the HH code can be dropped.